### PR TITLE
fix: improve last message performance (RC)

### PR DESCRIPTION
**Pull request by vitorhugods**
----
#### PR Submission Checklist for internal contributors

- The **PR Title**
  - [X] conforms to the style of semantic commits messages¹ supported in Wire's Github Workflow²
  - [X] answers the question: _If merged, this PR will: ..._ ³

- The **PR Description**
  - [X] is free of optional paragraphs and you have filled the relevant parts to the best of your ability
----

# What's new in this PR?

### Issues

Getting the last message of each conversation was SLOW when the Database was big.

### Causes

1. Inspecting the query, it seems that SQLite is Scanning the whole `MessageTextContent` table, and processing _all_ messages before grouping based on date.
2. We are missing DB indexes in multiple places
3. For a long conversation list, we're doing `O(N²)` in order to map conversations to last items

### Solutions

1. Change the query to first get the IDs of the wanted messages, and THEN select the `MessagePreviews` based on that.
On my DB (with 20k messages), after the Indexes were added, this change alone improved the query time on my desktop from 700ms to around 30ms. 

2. Add indexes for important columns, especially foreign keys. SQLite [does **NOT** create indexes automatically for foreign keys](https://www.sqlite.org/foreignkeys.html#fk_indexes).

3. This is a tiny optimisation, but transform the list of last messages to a map, for `O(1)` access time during the iteration later.

### Testing the Query on Desktop

Current query, no indexes (**_1088ms_**):

```sql
Execution finished without errors.
Result: 54 rows returned in 1088ms
At line 1:
SELECT * FROM MessagePreview AS message
WHERE visibility = 'VISIBLE' AND contentType IN ('TEXT', 'ASSET', 'KNOCK', 'MISSED_CALL', 'CONVERSATION_RENAMED', 'MEMBER_CHANGE')
GROUP BY conversationId
HAVING date = MAX(date)
ORDER BY message.date ASC;
```

New query, added indexes (**_85ms_**):
```sql
Execution finished without errors.
Result: 54 rows returned in 85ms
At line 1:
SELECT * FROM MessagePreview AS message
WHERE id IN (
	SELECT id FROM Message
	WHERE  
		Message.visibility = 'VISIBLE' AND 
		Message.content_type IN ('TEXT', 'ASSET', 'KNOCK', 'MISSED_CALL', 'CONVERSATION_RENAMED', 'MEMBER_CHANGE')
	GROUP BY Message.conversation_id
	HAVING Message.creation_date = MAX(Message.creation_date)
);

```

----
#### PR Post Merge Checklist for internal contributors

 - [X] If any soft of configuration variable was introduced by this PR, it has been added to the relevant documents and the CI jobs have been updated.
----
##### References
1. https://sparkbox.com/foundry/semantic_commit_messages
1. https://github.com/wireapp/.github#usage
1. E.g. `feat(conversation-list): Sort conversations by most emojis in the title #SQPIT-764`.

### CODE REVIEW

**Review by alexandreferris — APPROVED** [review]
queries be flyin 🚀

File: persistence/src/commonMain/db_user/com/wire/kalium/persistence/Messages.sq [hunk 1]
```diff
@@ -349,10 +355,14 @@ SELECT (
 
 getLastMessages:
 SELECT * FROM MessagePreview AS message
-WHERE visibility = 'VISIBLE' AND contentType IN ('TEXT', 'ASSET', 'KNOCK', 'MISSED_CALL', 'CONVERSATION_RENAMED', 'MEMBER_CHANGE')
-GROUP BY conversationId
-HAVING date = MAX(date)
-ORDER BY message.date ASC;
+WHERE id IN (
```
Comment (mchenani) [inline persistence/src/commonMain/db_user/com/wire/kalium/persistence/Messages.sq:358]: Nested queries are expensive than the joins, if we can convert that to a join that would be great!

**Review by mchenani — CHANGES_REQUESTED** [review]
Nice touch, 1 comment and 1 request:
yesterday I wrote some benchmarks to test the performance here, let's have some serious benchmark tests for this section as well. we can sync online about it.

**Review by Garzas — APPROVED** [review]
Unbreakable man 💪

File: persistence/src/commonMain/db_user/com/wire/kalium/persistence/Messages.sq (hunk 1 above)
Comment (vitorhugods) [reply persistence/src/commonMain/db_user/com/wire/kalium/persistence/Messages.sq:358]: I don't see a way to use a join for this 🤔 any tips?


Also, for what it's worth, there are no sub trees in the `EXPLAIN QUERY PLAN` for this one.

**Review by mchenani — APPROVED** [review]

**Pull request merged**